ocaml/xapi/xapi_vm_helpers.ml

```diff
@@ -722,16 +692,32 @@ let get_group_key ~__context ~vm =
   | Some x -> x
   | None -> `Other
 
-let group_hosts_by_best_pgpu ~__context vgpu =
-  let vgpu_type = Db.VGPU.get_type ~__context ~self:vgpu in
-  let gpu_group = Db.VGPU.get_GPU_group ~__context ~self:vgpu in
-  match
-    Xapi_gpu_group.get_remaining_capacity_internal ~__context
-      ~self:gpu_group ~vgpu_type
-  with
-  | Either.Left e -> raise e
-  | Either.Right _ -> ();
-    group_hosts_by_best_pgpu_in_group ~__context gpu_group vgpu_type
+(* Rank the provided hosts (visible_hosts) by the remainding capacity for the
+ * give vGPU. The vGPU has an associated vGPU group
+ * We get the rank preference from the vGPU grouop
+ * vgpu: the vGPU used to rank the provided hosts
+ * visible_hosts: the hosts will be ranked*)
+let rank_hosts_by_best_vgpu ~__context vgpu visible_hosts =
+  match visible_hosts with
+  | [] -> raise (Api_errors.Server_error (Api_errors.no_hosts_available, []))
+  | hosts ->
+    let vgpu_type = Db.VGPU.get_type ~__context ~self:vgpu in
+    let gpu_group = Db.VGPU.get_GPU_group ~__context ~self:vgpu in
+    let ordering =
+      match Db.GPU_group.get_allocation_algorithm ~__context ~self:gpu_group with
+      | `depth_first -> `ascending | `breadth_first -> `descending
+    in
+    Helpers.group_by ~ordering
+      (fun host ->
+         (* For every host
+          * 1. get all the pGPUs in the associated gpu_group
+          * 2. accumulate all the remaining capapcity for the vGPU
+          * 3. use the accumulated capacity as the ranking metrics*)
+         Db.Host.get_PGPUs ~__context ~self:host
+         |> Listext.List.intersect  (Db.GPU_group.get_PGPUs ~__context ~self:gpu_group)
```

Comment:
We can factor out `Db.GPU_group.get_PGPUs` and put it above the `group_by`, because the `gpu_group` doesn't change.